pkg/backends/calico/routes.go

```diff
@@ -46,15 +46,32 @@ type routeGenerator struct {
 	svcInformer, epInformer cache.Controller
 	svcIndexer, epIndexer   cache.Indexer
 	svcRouteMap             map[string][]string // maps service name to ip
+	clusterCIDR             string
 }
 
 // NewRouteGenerator initializes a kube-api client and the informers
-func NewRouteGenerator(c *client) (rg *routeGenerator, err error) {
+func NewRouteGenerator(c *client, clusterCIDR string) (rg *routeGenerator, err error) {
+	// Parse clusterCIDr to make sure it is valid.
+	cidr := strings.TrimSpace(clusterCIDR)
+	if _, _, err := net.ParseCIDR(cidr); err != nil {
+		return nil, fmt.Errorf("failed to parse cluster CIDR %s: %s", clusterCIDR, err)
+	}
+
+	// Determine the node name we'll use to check for local endpoints.
+	// This value should match the name of the node in the Kubernetes API.
+	// Prefer CALICO_K8S_NODE_REF, and fall back to the Calico node name.
+	nodename := template.NodeName
+	if n := os.Getenv("CALICO_K8S_NODE_REF"); n != "" {
+		nodename = n
+	}
+	log.Debugf("Route generator configured to use node name %s", nodename)
```

Comment:
This bit is something else I spotted - the NODENAME variable isn't always the same as the k8s node name.

So, we really should prefer the CALICO_K8S_NODE_REF variable, which is defined to the the name of the corresponding k8s node.

Reply:
(As long as this is only used for matching node name fields in the k8s API, which I think is the case.)